### DISCUSSION TITLE
Formatter Question for Multiline Strings

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#1526] [Q&A]
We have multiline strings in our model, like `terminal STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/;` (still in Langium 2.x).

We would like to have a formatter to take care of these multiline strings (e.g. removing any indentation):
```
person John "This is
             Multiline"

person Jane
"This is
Multiline (formatted)"

Hello John!
Hello Jane!
```

Is this somehow possible? What is the strategy, if so?

**Comment by msujew** [comment]
Hey @goto40,

given that Langium only formats the document content between two tokens (never within a token as that changes its semantic value), you will need to write the formatting code for that yourself. The main idea would be as follows:
1. Let the normal formatter run first
2. Find all multline string tokens in your AST (via `iterateCstTree` method available in the formatter)
3. Format those according to your liking by creating the appropriate text edits
4. Concat the results from (1) and (3) to create a final formatting

Note that I'm a bit concerned about the formatting behavior you're proposing. In most language, multi line strings aren't formatted in any way, because doing so will change the semantic content (i.e. the amount of whitespace within). Is that not of concern for your language?

**Reply by goto40** [reply]
Thank you again @msujew for this quick response!

I share your concern, but we have models with "descriptions" like in the following sketch - and we would also like to have homogeneously formatted models... (for code generation we already take care of the indentations).

Do you think we could do better somehow?

```
outer
"""
some description of outer
"""
{

  inner
  """
  some description of inner
  with indentation
  """
  {
  }

}
```

**Reply by msujew** [reply]
As long as whitespace within the multiline strings isn't semantically relevant (similar to markdown, where double whitespace gets rendered as a single whitespace), there's probably nothing to be concerned about. The 4 steps I've outlined above should do the trick.

**Reply by goto40** [reply]
Regarding your suggestion, on how to handle the multiline strings:

* Am I on the right track to override the `doDocumentFormat` method?
* If yes, I need some kind of "replace this text"-Action - are there some examples publicly available somewhere?
* I do not really understand the data flow concerning the `context`: do I have the chance to get the indentation level I defined in the original formatter somehow?

```javascript
    protected override doDocumentFormat(document: LangiumDocument, options: FormattingOptions, range?: Range): TextEdit[] {
        // (1) prepare original formatting
        const originalEdits = super.doDocumentFormat(document, options, range);

        // (2), (3) find all multiline tokens (`Person: ... info=STRING;`)
        const additionalEdits = new Array<TextEdit>();
        const context: FormattingContext = {
            indentation: 0,
            options,
            document: document.textDocument
        };
        this.iterateCstTree(document, context).forEach(node=>{
            if (ast.isPerson(node.astNode)) {
                const assignment = GrammarUtils.findAssignment(node);
                if (assignment) {
                    if (assignment.feature==='info') {
                        //const edits = TODO create edits here for the assignment value
                        //additionalEdits.push(...edits);
                    }
                }
            }
        });

        // (4) merge the lists (do I need to call avoidOverlappingEdits)
        return originalEdits.concat(additionalEdits);
    }
```
